### ISSUE TITLE
Mutate workflow namespaces

### ISSUE THREAD

**Issue by denis-yuen** [ga4gh/dockstore#1645]
### Desired behaviour

As discussed in a developer biweekly, need to mutate hosted workflow name generation to avoid email leakage.

**Comment by agduncan94** [comment]
What was the decision for this?

**Comment by denis-yuen** [comment]
Look for "Bi-Weekly Dockstore Developers Meeting" in the shared collaboration folder
```
Discuss naming for hosted dockstore workflows
Are users comfortable with workflows including their emails?
How to we achieve unique naming across email providers and github without exposing emails?
Is censoring ala google groups a possibility?
Can be change the naming scheme trivially?
Examples:
dockstore.org/agduncan94/bamqc
dockstore.org/<new signup names>/bamqc

(decision below)
dockstore.org/agduncan:goo.gl/bamqc
dockstore.org/agduncan(:github.com)/bamqc
```

**Comment by denis-yuen** [comment]
Alternatives to `goo.gl ` welcome, @agduncan94 @coverbeck  but the idea was to make it non-trivial to harvest email addresses from our APIs

**Comment by coverbeck** [comment]
I forget, did we decide what would happen if two agduncans from different domains signed up? How would the `goo.gl` be different for the second one in this proposal?

I don't like `goo.gl` that much, because for some reason the `goo` sticks out to me, but before proposing an alternative, I'm wondering what the rules are.

**Comment by denis-yuen** [comment]
If I understand correctly, are you asking if two agduncans from domains that use google login sign up? 
I guess I was assuming that only gmail logins would work, but I'm seeing from your question that other google domains might need to be considered (is this for Google Apps users? Not very familiar with how this works)

**Comment by coverbeck** [comment]
Yes, that's what I'm wondering about. For example, I can now log into Dockstore via Google with my ucsc.edu account; another user with my name could also log in from a gmail.com account.

**Comment by denis-yuen** [comment]
Ah, cool/rats. 
I didn't realize that that would now work, but that does complicate the naming issue.

**Comment by denis-yuen** [comment]
As discussed in biweekly, one solution to try out. 

0) Add a constraint for unique usernames
1) Login
  a) With Github, you get your Github username with no changes. The change is when a name clash occurs, mutate the username by mutation (counting up is one option)
  b) With Google, we use the email as we currently do with no changes 
  c) In both cases, on first login (see 1 below), if you haven't previously changed your username you can change your username by web service call. 
2) After creating tools and workflows, you will no longer be able to change your username 

Changes required: 
1) Database constraint, one column to check whether a name change has occurred/name has been confirmed
2) New web service call to change username (if you are logged in as that user)
3) When logging in with a name that has not changed, you will be prompted to optionally change your name or confirm your name
4) When logging in with a confirmed name (or one that has content if they created content via API or bypassed name change screen), you will login normally.

**Comment by denis-yuen** [comment]
I think this has basically changed to a user friendliness and documentation ticket. 
1. Try it out as a google login, see whether you can change your username and create workflows
2. Try out creating workflows without changing your name and self destructing and trying again
3. If anything is not intuitive make ui changes, also highly encourage Google registrations to change their username to avoid exposing emails

**Comment by agduncan94** [comment]
For 1.
I changed my Google username, however when I created a hosted workflow it used my original email username. Logging in and out does not solve this. My username is properly updated in the database.

It seems like the user object we are passing the create hosted still has the old username. A quick fix is to user the id of the logged in user to grab the updated user object from the database in the webservice.

For 2.
Doesn't seem like the workflow is deleted.

> [HTTP 400] Bad Request: A workflow already exists with that path. Please change the workflow name to something unique.

For 3.
I will create a PR
